### ISSUE TITLE
Scalar variables for 1-sample delay lines

### ISSUE THREAD

**Issue by mikeoliphant** [grame-cncm/faust#567]
Because single-sample feedback is so common, Faust often generates a lot of 2-value arrays for feedback handling.

This ends up being quite expensive in languages like C# where array access has more overhead than C++.

It would be great if the compiler could instead generate a pair of scalar variables for the single-sample delay case.

ie: instead of:

`double[] fRec9 = new double[2];`

it would generate something like:

`double fRec9_0; double fRec9_1`

In C#, this simple change can result in massive performance increases for very typical Faust use cases. It doesn't seem to make much of a difference for C++, but it may help other backends.

**Comment by sletz** [comment]
Yes indeed. Yann is currently working on a whole new infrastructure in the compiler that will (in particular but even some more improvements ...) allow to solve this kind of issues. See Yann IFC 2020 presentation: https://ifc20.sciencesconf.org and https://www.youtube.com/watch?v=vb7LU7ue5hk.

**Comment by mikeoliphant** [comment]
Great - I look forward to the new compiler!

**Comment by mikeoliphant** [comment]
Hi @sletz - my main use case was simply to be able to generate plugins in C# rather than having to use C#/C++ interop (which isn't difficult, but is unpleasant and make for more complicated project structure). I also am using the ability of the C# compiler to do dynamic assembly generation from Faust code. One goal of this would be to have runtime editing of Faust code within a VST inside a DAW (something I've experimented with, but haven't had time to fully explore).

**Comment by sletz** [comment]
Thanks. How do you do use the C# compiler to do dynamic assembly generation from Faust code ?

**Comment by sletz** [comment]
OK thanks, yes using LLVM backend, we do DSP ==> LLVM IR ==> JIT ==> executable code in memory.

**Comment by mikeoliphant** [comment]
Hi @sletz - figured I'd contact you here because of our previous discussion.

I've *finally* gotten around to making more progress on one of my use cases for the C# faust compiler.

I just released the first version of:

https://github.com/mikeoliphant/FaustVst

It dynamically compiles faust dsp code into a C# assembly while running inside a VST plugin. It makes for a very good workflow for quick iteration and testing inside a DAW.

**Comment by sletz** [comment]
Thanks, added [here](https://faust.grame.fr/community/powered-by-faust/#faustvst).

**Comment by mikeoliphant** [comment]
👍

**Comment by expertsleepers** [comment]
I just came here to request the same thing, but with a different motivation.

I'm working on a toolchain to get Faust code running on an embedded platform:
https://github.com/expertsleepersltd/distingNT_API/tree/main/faust

As such, it has various forms of memory - mainly large off-chip DRAM, and small but fast on-chip RAM. For this reason I have to use the -mem option so that I can separate out the larger delay buffers etc. into DRAM. However, because of the issue mentioned above, all my single sample delay lines are also going into DRAM, which is slower anyway, and bloats the code because there's an extra pointer read to access them.

My suggested solution was going to be another compiler option so you can specify a threshold below which the memory manager is not used. So you could say, please don't use an allocated array for an array of two floats, just put it directly inside mydsp as if -mem was not used.

I've yet to dig into the faust codebase. If it doesn't seem easy to do, I'll likely do this as a post-process on the C++ files in a Python script.

**Comment by sletz** [comment]
Interesting, it would help if you can paste here a simple Faust DSP showing the issue, and explaining what you would like to see changed in the generated code.

**Comment by expertsleepers** [comment]
Sure.

```
import("stdfaust.lib");

process = dm.greyhole_demo;
```

```
struct mydsp final : public dsp {
	
	FAUSTFLOAT fHslider0;
	float* fVec0;
	int* iVec1;
	FAUSTFLOAT fHslider1;
	float* fRec6;
	FAUSTFLOAT fHslider2;
	float* fVec2;
	float* fRec10;
	float* fRec14;
	float* fRec21;
	float* fRec25;
	float* fRec29;
	float* fRec36;
	float* fRec40;
	float* fRec44;
	FAUSTFLOAT fHslider3;
	float* fVec3;
	int fSampleRate;
	float fConst0;
	float fConst1;
	FAUSTFLOAT fHslider4;
	float* fVec4;
	float fConst2;
	FAUSTFLOAT fHslider5;
	float* fVec5;
	float* fRec48;
	float* fRec49;
	int IOTA0;
	float* fVec6;
```

```
	void memoryCreate() {
		fVec0 = static_cast<float*>(fManager->allocate(8));
		iVec1 = static_cast<int*>(fManager->allocate(8));
		fRec6 = static_cast<float*>(fManager->allocate(8));
		fVec2 = static_cast<float*>(fManager->allocate(8));
		fRec10 = static_cast<float*>(fManager->allocate(8));
		fRec14 = static_cast<float*>(fManager->allocate(8));
		fRec21 = static_cast<float*>(fManager->allocate(8));
		fRec25 = static_cast<float*>(fManager->allocate(8));
		fRec29 = static_cast<float*>(fManager->allocate(8));
		fRec36 = static_cast<float*>(fManager->allocate(8));
		fRec40 = static_cast<float*>(fManager->allocate(8));
		fRec44 = static_cast<float*>(fManager->allocate(8));
		fVec3 = static_cast<float*>(fManager->allocate(8));
		fVec4 = static_cast<float*>(fManager->allocate(8));
		fVec5 = static_cast<float*>(fManager->allocate(8));
		fRec48 = static_cast<float*>(fManager->allocate(8));
		fRec49 = static_cast<float*>(fManager->allocate(8));
		fVec6 = static_cast<float*>(fManager->allocate(524288));
```

(and there are many, many more 8 byte allocations)

Everything there apart from fVec6 could just be e.g.

```
	float fVec0[2];
```

**Comment by sletz** [comment]
OK thanks. TBH right now  the `-memX` related code is a bit of a mess... ((-; and I don't have much time now to rework all that in a more flexible way. Two things:

- event if small arrays are seen as pointer and handled by the memory manager, since you know their size why can't you allocate them in on-chip RAM ?
-  if we design a new options, what would be more appropriate: a size threshold for each individual array ? Or a global "on-chip RAM" memory size option and the DSP array would be allocate in the DSP structure until the global "on-chip RAM" memory size would be exhausted ? Or a mix of the 2 options.

Anyway I guess the way to move on for now is probably the Python post-process as you suggested. And I'll follow the distingNT_API and am interested to see how Faust generally behave there. Ans you may be interested by [optimisation general advices](https://faustdoc.grame.fr/manual/optimizing/), especially useful on embedded use-case.

**Comment by expertsleepers** [comment]
- Yes, that is an option (to allocate into different areas depending on size). It was what I first considered doing. But it would generate smaller and more efficient code to have them as fixed arrays in mydsp.
- I think it would be fine to have a global heuristic for allocated vs non-allocated arrays, but I admit I'm not speaking from a position of a great deal of experience.

I don't actually now how widespread this kind of small array is in Faust code in general. It could be that directly changing the code for the single sample delay case, as suggested in the OP, actually catches 99.9% of the relevant cases.

Another data point - in that greyhole demo, the compute() call compiles down to about 6kb on the target. memoryInfo() and memoryCreate() together generate about 3kb of code, because of all the tiny allocations. When code memory is tightly constrained, this is also an issue.

**Comment by expertsleepers** [comment]
There's also the very general consideration that using memory to store a pointer to a small allocation is very wasteful of memory. On a 64 bit platform using floats, it's using an 8 byte pointer to point to an 8 byte allocation - which could simply go in the memory used for the pointer!

**Comment by sletz** [comment]
Several comments here:

- we also are in the process of implementing the new `ondemand` primitive, and we are also reworking the delay (and recursion) related delay lines implementations. This is WIP but already visible here: https://github.com/grame-cncm/faust/tree/master-dev-ocpp-od-fir-2-FIR. All of the changes are in the `-lang cpp` backend. 

- but even on this branch the "Scalar variables for 1-sample delay lines (or recursion BTW...)" optimisation is not there. @orlarey : so the embedded use-case here is a real motivation to have the optimisation, we should do that !

- so @expertsleepers we'll need a bit of time to sort all that and hopefully improve the generated code further to answer to your needs

**Comment by expertsleepers** [comment]
I wrote my Python script to remove the small arrays (which is now in my GitHub), and I have some concrete results, based on the greyhole.dsp demo.

Before:
compute() 8976 bytes
memoryCreate() 1204 bytes
CPU: 34%

After:
compute() 8008 bytes (about 11% smaller)
memoryCreate() 280 bytes
CPU: 28%

What we don't know from that is how much of the CPU saving is down to moving the small data into fast RAM vs the code being more efficient without the pointer indirection. But even ignoring the CPU saving, the code memory saving is very much worth having.

**Comment by sletz** [comment]
Very interesting thanks, so even more justification to do that in the compiler.
